[FSDP2] allow meta tensors during loading state dict and cpu offloading

weifengpy opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126305
* __->__ #126267

unit test: ``pytest test/distributed/_composable/fsdp/test_fully_shard_state_dict.py``

with meta init and cpu offloading, we have meta tensors after`model.load_state_dict(assign=True, strict=False)`. This PR avoided calling `.cpu` on meta tensors otherwise it's a runtime error

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

weifengpy committed:
[FSDP2] allow meta tensors during loading state dict and cpu offloading

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

awgu commented on test/distributed/_composable/fsdp/test_fully_shard_state_dict.py:
```diff
@@ -68,6 +68,42 @@ def _test_1d_state_dict_save_load(self, mlp_dim: int):
+            )
+            state_dicts.append({name: sharded_tensor})
+
+        # allow meta params during `reset_sharded_param`
```
nit: maybe something like this?
```suggestion
        # check that we can load with some parameters still on meta device
```
`reset_sharded_param` is an implementation detail (and may be refactored without remembering to change this comment, e.g. if we were to remove the method altogether after padding local tensor)

awgu commented on test/distributed/_composable/fsdp/test_fully_shard_state_dict.py:
```diff
@@ -68,6 +68,42 @@ def _test_1d_state_dict_save_load(self, mlp_dim: int):
+        inp = torch.rand((mlp_dim, mlp_dim), device="cuda")
+        model(inp)
+
+        state_dicts = model.state_dict()
```
nit: this one seems should just be one state dict (singular, not plural)
```suggestion
        state_dict = model.state_dict()
```

awgu reviewed (approved):
LGTM! I left some nits on the unit test.

weifengpy committed:
Update on "[FSDP2] allow meta tensors during loading state dict and cpu offloading"


unit test: ``pytest test/distributed/_composable/fsdp/test_fully_shard_state_dict.py``

with meta init and cpu offloading, we have meta tensors after`model.load_state_dict(assign=True, strict=False)`. This PR avoided calling `.cpu` on meta tensors otherwise it's a runtime error

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

weifengpy commented:
@pytorchmergebot merge

awgu commented on torch/distributed/_composable/fsdp/_fsdp_param.py:
```diff
@@ -245,7 +245,7 @@ def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
         if sharded_param.numel() > 0:
             padded_sharded_param[: sharded_param.size(0)].copy_(sharded_param)
-        if self.offload_to_cpu:
+        if self.offload_to_cpu and not torch.empty(0).is_meta:
```
```suggestion
        if self.offload_to_cpu and not padded_sharded_param.is_meta:
```

weifengpy replied:
resolving in https://github.com/pytorch/pytorch/pull/126305
